### PR TITLE
mipsel: switch to n64 abi

### PR DESCRIPTION
**Pull request by FlyGoat**
mipsel: switch to n64 abi

### CODE REVIEW

**Comment by FlyGoat** [comment]
After discussion, we decided to set MIPS-III N64 ABI as mips64el mips64el-aosc-linux-gnu, and MIPS64r2 N64 ABI as mips64r2el  mipsisa64r2el-aosc-linux-gnu.

**Comment by lmy441900** [comment]
Developers:

At this time please do not merge, to allow our thorough research on these changes.

**Comment by Artoria2e5** [comment]
yo merge #102 into your thing

File: arch/_common.sh [hunk 1]
```diff
@@ -8,14 +8,14 @@ ARCH_TARGET['amd64']=x86_64-aosc-linux-gnu
 ARCH_TARGET['armel']=armv7a-aosc-linux-gnueabihf
 ARCH_TARGET['i386']=i686-aosc-linux-gnu
 ARCH_TARGET['mips64el']=mips64el-aosc-linux-gnu
```
Comment (lmy441900) [inline arch/_common.sh]: Actually I propose `mips64el-aosc-linux-gnuabin64` here if we are going to have o32 and n32 compilers.

File: arch/_common.sh [hunk 2]
```diff
@@ -8,14 +8,14 @@ ARCH_TARGET['amd64']=x86_64-aosc-linux-gnu
...
-ARCH_TARGET['mipsel']=mipsel-aosc-linux-gnu
+ARCH_TARGET['mips64r2el']=mipsisa64r2el-aosc-linux-gnu
```
Comment (lmy441900) [inline arch/_common.sh]: `mipsisa64r2el-aosc-linux-gnuabin64` here too (I admit that is too long).

File: arch/mips64el.sh [hunk 3]
```diff
@@ -1,11 +1,9 @@
...
-#   - Artoria Wang <arthur2e5@aosc.xyz>
-#   - Bob Cao <bobcao3@aosc.xyz>
-#   - Mingcong Bai <jeffbai@aosc.xyz>
-
-CFLAGS_COMMON_ARCH=' -mabi=64 -march=mips64r2 -mtune=loongson3a '
+# FIXME: Loongson workaround added as suggested by biergaizi, do this break anything?
+# Global -mshared flag since GCC sets -mno-shared by default when compiling non-PIC.
+# Not using -fPIC globally to avoid overriding -fPIE in GCC specs.
+# Adapted from Void Linux: https://github.com/voidlinux/void-packages/commit/882f23cf980d2277594b5f5c036114c0c4d9228a
+CFLAGS_COMMON_ARCH='-march=mips2 -mtune=loongson2f -mabi=64 -Wa,-mfix-loongson2f-nop -mshared '
```
Comment (lmy441900) [inline arch/mips64el.sh]: `-march=mips3`

More flags...? (We may fix more)

**Comment by Artoria2e5** [comment]
pkgprovide exists. you can have a new proc or acbs wrap handle this overlay shit. maybe work it into the new arch system to-do too.

**Comment by MingcongBai** [comment]
@Artoria2e5 The overlay system utilises repositories, not so much on a packaging level, the triplet wouldn't change, only the flags.

Aside: please keep it civilised. As fun as it is to speak your true mind, it is much easier to engage in a technical conversation when we could all speak respectfully.

File: arch/mips64r2el.sh [hunk 4]
```diff
@@ -8,4 +8,4 @@
 #   - Bob Cao <bobcao3@aosc.xyz>
 #   - Mingcong Bai <jeffbai@aosc.xyz>
 
-CFLAGS_COMMON_ARCH=' -mabi=64 -march=mips64r2 -mtune=loongson3a '
+CFLAGS_COMMON_ARCH=' -mabi=64 -march=mips64r2 -mtune=loongson3a -mfix-loongson3-llsc -mfx-loongson3-loads'
```
Comment (lmy441900) [inline arch/mips64r2el.sh]: > `-mfx-loongson3-loads`

Typo + loss of space at the end

**Comment by FlyGoat** [comment]
#101 Now at least MIPS64r2EL works.
Requst for some review.
I'm going to merge it later.

File: arch/mips64r2el.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+# Suggested, discussed by:
+#   - Junde Yhi <lmy441900@aosc.xyz>
+#   - Artoria Wang <arthur2e5@aosc.xyz>
+#   - Bob Cao <bobcao3@aosc.xyz>
+#   - Mingcong Bai <jeffbai@aosc.xyz>
+# Switched to mips64r2el by:
+#   - Jiaxun Yang <jiaxun.yang@flygoat.com>
+
+
+CFLAGS_COMMON_ARCH='-EL  -mabi=64 -march=mips64r2 -mtune=loongson3a  -mllsc -mno-madd4'
```
Comment (lmy441900) [inline arch/mips64r2el.sh]: Is `-EL` really necessary?

File: arch/mips64r2el.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+# Loongson binutils fixes -mfix-loongson3-llsc, -mfix-loongson3-load is disableed. And should
+# be enabled in binutils patch.
```
Comment (lmy441900) [inline arch/mips64r2el.sh]: `disabled`.

I do not get your idea. Do you mean we should only use `-mfix-loongson3-load` in binutils but not globally?

**Comment by lmy441900** [comment]
> I'm going to merge it later.

Not now.

**Comment by Artoria2e5** [comment]
I have no idea about what I am reading right now, and I don't think you should ask me for a review.

**Comment by lmy441900** [comment]
> and I don't think you should ask me for a review.

Whether the code quality is good or not.

File: etc/autobuild/ab3_defcfg.sh [hunk 7]
```diff
@@ -126,7 +126,9 @@ abdetectarch(){
 		aarch64) echo arm64 ;;
 		ppc) echo powerpc ;;
 		ppc64) echo ppc64 ;;
-		mips64) echo mips64el ;;
+    mips64) if [[ "$(readelf -h /bin/bash | grep -c "mips3")" == "0" ]]
```
Comment (MingcongBai) [inline etc/autobuild/ab3_defcfg.sh]: Mind the alignment here.

**Comment by MingcongBai** [comment]
Looks good to me for the most part. Though mind the alignment in case `mips64)` from `ab3_defcfg.sh`.

**Comment by MingcongBai** [comment]
@FlyGoat Please respond to some of the review questions above...

**Comment by FlyGoat** [comment]
@MingcongBai ok I'm going to fix it tomorrow.(strange behavior of vscode....

**Comment by lmy441900** [comment]
@FlyGoat Do you really know what are you doing?

**Comment by lmy441900** [comment]
You should really care about people's opinions, since you are not working on a personal project.

In eb5d8ae you did add code to support `mipsisa64r2el`, but you 1) did not respond questions above, 2) bypassed this PR, rendering workflow messy.

You should think about it. This PR is not fully resolved, reopening.

**Comment by FlyGoat** [comment]
@lmy441900 my fault.... Just trying to find out a method to separate changes between mips64el and mips64el since origin commits mixed them together.

**Comment by FlyGoat** [comment]
@lmy441900 Should I create another pr for mips64el or continue this PR?

**Comment by lmy441900** [comment]
And now we can see conflicts.

You do great on porting AOSC OS to MIPS (Loongson) platforms. You've made a leap. But think about quality of code. Your production should be easy for other people to maintain (understand).

You can of course continue to push to staging, but please leave this PR open as an issue to keep track of the status of support.

File: arch/mips64el.sh [hunk 8]
```diff
@@ -1,11 +1,9 @@
 #!/bin/bash
-##arch/mips64el.sh: Build definitions for mip64el (optimized for Loongson 3A).
+##arch/mipsel.sh: Build definitions for mipsel (optimized for Loongson 2F).
```
Comment (Yexiaoxing) [inline arch/mips64el.sh]: If you change this comment, remember to change the file name.

Comment (MingcongBai) [reply arch/mips64el.sh]: @FlyGoat Please take note...

**Comment by MingcongBai** [comment]
@FlyGoat Please limit your work to what this PR is about. What we needed to do here is to add an architectural support, no more, no less. If a function (related or not) needs rework, please open up another PR - this PR is big enough as is...

**Pull request closed without merging**